country codes are missing for some entries.

In [1]:
import csv

##### Task 1

In [2]:
data = list(csv.reader(open('Summer_olympic_Medals.csv','r')))

In [3]:
data[0:5]

[['Year',
  'Host_country',
  'Host_city',
  'Country_Name',
  'Country_Code',
  'Gold',
  'Silver',
  'Bronze'],
 ['1896', 'Greece', 'Athens', 'Great Britain', 'GBR', '2', '3', '2'],
 ['1896', 'Greece', 'Athens', 'Hungary', 'HUN', '2', '1', '3'],
 ['1896', 'Greece', 'Athens', 'France', 'FRA', '5', '4', '2'],
 ['1896', 'Greece', 'Athens', 'United States', 'USA', '11', '7', '2']]

##### Task 2

In [4]:
header = data.pop(0)

In [5]:
header

['Year',
 'Host_country',
 'Host_city',
 'Country_Name',
 'Country_Code',
 'Gold',
 'Silver',
 'Bronze']

In [6]:
len(data)

1344

There are 1344 entries plus a header (column names) in the data.

#### Tasks 3 to 10

Structure for storing countries, years, medals: {country: [[year1, golds1, silvers1, brozes1], ...[yearN, goldsN, silversN, brozesN]]}

In [7]:
data_dict = {}
country_codes = {}

for y_c in data:  
    # loop through the data as unique year and country combinations
    codes_countries = {country:code for code, country in country_codes.items()}
    if y_c[4]=='':
        try:
            y_c[4] = codes_countries[y_c[3]]
        except:
            y_c[4] = "_".join(y_c[3].split(' '))
    if y_c[4] in data_dict:
        data_dict[y_c[4]] = data_dict[y_c[4]] + [[y_c[0], y_c[5], y_c[6], y_c[7]]]
    else: 
        country_codes[y_c[4]]=y_c[3] # correspond country codes with full name
        data_dict[y_c[4]] = [[y_c[0], y_c[5], y_c[6], y_c[7]]]

In [8]:
data_dict = {c: [list(map(int,val)) for val in y_c] for c,y_c in data_dict.items()}

##### Task 3

Instruction: compute the average numbers of gold, silver, and bronze medals per country and write this (in decreasing order) to a file ”average medals per country.csv”

Compute total medals each year

In [9]:
data_totals_dict = {c: [y_c + [sum(y_c[1:])] for y_c in data_dict[c]] for c in data_dict}

In [10]:
avg_medals = {}
avg_medals = {c: list(map(lambda x: x/len(data_totals_dict[c]), # averaged over years
     [sum([y[1] for y in  data_totals_dict[c]]), # total of golds
     sum([y[2] for y in  data_totals_dict[c]]), # silvers
     sum([y[3] for y in  data_totals_dict[c]]),  # bronzes
     sum([y[4] for y in  data_totals_dict[c]])])) # all medals
     for c in data_totals_dict} # averaged

In [11]:
avg_medals_list = [['Country', 'Avg_Gold_Medals','Avg_Silver_Medals',
                'Avg_Bronze_Medals','Avg_Total_Medals']] + \
[[country_codes[c[0]]]+ c[1] for c in list(dict(sorted(
    {c:m for c,m in avg_medals.items()}.items(), 
    key = lambda x: x[1][3], reverse=True)).items())
]

In [12]:
with open('average_medals_per_country.csv','w') as medals_csv:
    wr = csv.writer(medals_csv)
    for row in avg_medals_list:
        wr.writerow(row)
    medals_csv.close()
# medals_csv = open('average_medals_per_country.csv','r')
# medals_csv.read()

##### Highest Average Medals

In [80]:
avg_medals_list[1]

['Soviet Union',
 43.888888888888886,
 35.44444444444444,
 32.888888888888886,
 112.22222222222223]

Highest avg medals: Soviet Union

##### 5. Top 10 countries by mean number of (total) medals

In [82]:
top_ten_avg_medals = [c[0] for c in medals_list[1:11]]

In [20]:
top_ten_avg_medals

['Soviet Union',
 'Unified Team',
 'United States',
 'East Germany',
 'Russia',
 'China',
 'West Germany',
 'United Team of Germany',
 'Germany',
 'Great Britain']

['Soviet Union',
 'Unified Team',
 'United States',
 'East Germany',
 'Russia',
 'China',
 'West Germany',
 'United Team of Germany',
 'Germany',
 'Great Britain'] have the highest average medals.

##### 6.  compute the median number of gold, silver, and bronze medals per country and write this (in decreasing order) to a file "median_medals_ per_country.csv"

In [21]:
def median(numbers):
    '''take a list or tuple of numbers and return the median'''
    size = len(numbers)
    numbers = sorted(numbers)
    if size == 1:
        med = numbers[0]
    elif size%2 == 0:
        med = (numbers[int(size/2-1)]+numbers[int(size/2)])/2
    else:
        med = numbers[size//2]
    return med

In [22]:
med_medals_dict = {c:[median([y[1] for y in data_totals_dict[c]])] + \
                   [median([y[2] for y in data_totals_dict[c]])] + \
                   [median([y[3] for y in data_totals_dict[c]])] + \
                   [median([y[4] for y in data_totals_dict[c]])]
                   for c in data_totals_dict.keys()}

In [23]:
median_medals = dict(sorted(med_medals_dict.items(), key = lambda x: x[1][3], reverse=True))
median_medals

{'EUN': [45, 38, 29, 112],
 'URS': [43, 31, 32, 99],
 'USA': [36.5, 27.5, 27.0, 94.0],
 'GDR': [37, 25, 25, 90],
 'RUS': [25.0, 20.5, 25.5, 65.5],
 'CHN': [27.0, 20.0, 15.0, 60.5],
 'EUA': [10, 19, 11, 42],
 'FRG': [11, 12, 16, 40],
 'GER': [10, 11, 14, 37],
 'GBR': [5, 9, 9, 23],
 'ITA': [8, 8, 8, 23],
 'UKR': [5, 5, 10, 22],
 'JPN': [6.5, 7.0, 7.5, 20.5],
 'KOR': [6, 4, 7, 20],
 'FRA': [7, 6, 7, 19],
 'HUN': [7, 5, 5, 18],
 'AUS': [5, 6, 5, 17],
 'CUB': [5.0, 3.5, 4.5, 14.0],
 'ROU': [4.0, 4.5, 6.0, 14.0],
 'BLR': [2, 4, 6, 13],
 'POL': [3.0, 4.5, 5.0, 12.5],
 'SWE': [4, 5, 6, 12],
 'NED': [3.5, 3.5, 4.0, 10.5],
 'CAN': [3, 3, 5, 10],
 'BUL': [3, 4, 3, 10],
 'CZE': [3, 3, 4, 10],
 'KEN': [2, 4, 2, 9],
 'KAZ': [3, 4, 4, 9],
 'TCH': [2.5, 3.0, 2.5, 8.0],
 'DEN': [2.0, 2.0, 3.0, 6.0],
 'ANZ': [1.5, 2.0, 2.5, 6.0],
 'GEO': [2, 2, 2, 6],
 'AZE': [1, 1, 4, 6],
 'SRB': [1.5, 1.0, 2.0, 6.0],
 'SUI': [1.0, 2.0, 2.0, 5.5],
 'FIN': [3.0, 2.0, 3.0, 5.5],
 'TUR': [2.0, 1.0, 1.5, 5.5],
 'NOR': [2,

In [24]:
med_medals_list = [['Country', 'Med_Gold_Medals','Med_Silver_Medals',
                'Med_Bronze_Medals','Med_Total_Medals']] + \
[[country_codes[c[0]]]+ c[1] for c in list(dict(sorted(
    {c:m for c,m in median_medals.items()}.items(), 
    key = lambda x: x[1][3], reverse=True)).items())
]

In [25]:
with open('median_medals_per_country.csv','w') as median_medals_csv:
    wr = csv.writer(median_medals_csv)
    for row in med_medals_list:
        wr.writerow(row)
    median_medals_csv.close()
# median_medals_csv = open('median_medals_per_country.csv','r')
# median_medals_csv.read()

##### 7. Which country has the highest median number of medals?

In [26]:
med_medals_list[1][0]

'Unified Team'

Q7 Answer: Unified Team has the highest median number of medals.

##### 8. list top 10 countries by median number of medals

In [27]:
[c[0] for c in med_medals_list[1:11]]

['Unified Team',
 'Soviet Union',
 'United States',
 'East Germany',
 'Russia',
 'China',
 'United Team of Germany',
 'West Germany',
 'Germany',
 'Great Britain']

Q8 answer: 'Unified Team',
 'Soviet Union',
 'United States',
 'East Germany',
 'Russia',
 'China',
 'United Team of Germany',
 'West Germany',
 'Germany',
 'Great Britain'

##### 9. For which countries you have #(gold) = #(silver) = #(bronze). For which countries you have (#(gold) + #(silver) ) = #(bronze)

###### 9a: which countries have the same numbers of gold, silver, and bronze medals.

In [28]:
# make a dictionary that corresponds the country with their aggregated totals for 
# gold, silver, bronze, and overall accross all years

medal_totals = {c:  # averaged over years
     [sum([y[1] for y in  data_totals_dict[c]]), # total of golds
     sum([y[2] for y in  data_totals_dict[c]]), # silvers
     sum([y[3] for y in  data_totals_dict[c]]),  # bronzes
     sum([y[4] for y in  data_totals_dict[c]])] # all medals
     for c in data_totals_dict} # averaged

In [29]:
{c:m for c,m in medal_totals.items() if (m[0]==m[1])&(m[1]==m[2])}

{'TTO': [1, 1, 1, 3], 'GRN': [1, 1, 1, 3]}

In [30]:
[country_codes[c] for c in {c:m for c,m in medal_totals.items() if (m[0]==m[1])&(m[1]==m[2])}.keys()]

['Trinidad and Tobago', 'Grenada']

9a: Trinidad and Tobago along with Grenada are the two countries that have equal numbers of gold, silver and bronze medals. 

###### 9b: Which countries have the same number of bronze medals as the sum of gold and silver medals.

In [31]:
[country_codes[c] for c in 
 {c:m for c,m in medal_totals.items() if (m[0]+m[1])==m[2]}]

['Haiti',
 'Portugal',
 'Egypt',
 'Lebanon',
 'Iceland',
 'Morocco',
 'Niger',
 'Bermuda',
 'Zambia',
 'Syria',
 'Ivory Coast',
 'Chinese Taipei',
 'Suriname',
 'Costa Rica',
 'Lithuania',
 'Kazakhstan',
 'Mozambique',
 'Saudi Arabia',
 'Macedonia',
 'United Arab Emirates',
 'Tajikistan',
 'Botswana',
 'Independent Olympic Athletes',
 'Jordan',
 'Fiji']

9b Answer: 'Haiti',
 'Portugal',
 'Egypt',
 'Lebanon',
 'Iceland',
 'Morocco',
 'Niger',
 'Bermuda',
 'Zambia',
 'Syria',
 'Ivory Coast',
 'Chinese Taipei',
 'Suriname',
 'Costa Rica',
 'Lithuania',
 'Kazakhstan',
 'Mozambique',
 'Saudi Arabia',
 'Macedonia',
 'United Arab Emirates',
 'Tajikistan',
 'Botswana',
 'Independent Olympic Athletes',
 'Jordan',
 'Fiji'

##### 10. Which countries increased their number of total medals each game during the last 4 Olympic games and which coun- tries decreased their total number of medals during the last 4 Olympic games?

In [45]:
# Determine who participated in the last 5 Olympics
def participated_l_5(d = dict()):
    '''take a dictionary of coutry:[list of years and medals] and determine if 
    they competed in all of the last five Olympics, 
    (2000, 2004, 2008, 2012, and 2020), returning a revised dictionary with countries
    remaining if they satisfy this condition.'''
    participated = {}
    for country, year_medals in d.items():
        if [y[0] for y in year_medals[-5:]] == \
        [2004, 2008, 2012, 2016, 2020]:
            participated[country] = True
        else:
            participated[country] = False
    return {c:d[c] for c,p in participated.items() if p == True}

###### 10a Increasing medals, i.e. 2020<2016<2012<2008<2004.

In [68]:
def increasing_medals(d = dict()):
    '''return a dictionary including the items for which there is an increase
    in total medals for each of the last four olympics'''
    return {c: y_m for c, y_m in d.items() if \
                 (y_m[-1][4] > y_m[-2][4])&\
                 (y_m[-2][4] > y_m[-3][4])&\
                 (y_m[-3][4] > y_m[-4][4])\
                 &(y_m[-4][4] > y_m[-5][4])
           }

In [46]:
participated_l_5(data_totals_dict)

{'GBR': [[1896, 2, 3, 2, 7],
  [1900, 15, 8, 9, 32],
  [1904, 1, 1, 0, 2],
  [1908, 56, 51, 39, 146],
  [1912, 10, 15, 16, 41],
  [1920, 14, 16, 13, 43],
  [1924, 9, 13, 12, 34],
  [1928, 3, 10, 7, 20],
  [1932, 4, 7, 5, 16],
  [1936, 4, 7, 3, 14],
  [1948, 3, 14, 6, 23],
  [1952, 1, 2, 8, 11],
  [1956, 6, 7, 11, 24],
  [1960, 2, 6, 12, 20],
  [1964, 4, 12, 2, 18],
  [1968, 5, 5, 3, 13],
  [1972, 4, 5, 9, 18],
  [1976, 3, 5, 5, 13],
  [1980, 5, 7, 9, 21],
  [1984, 5, 11, 21, 37],
  [1988, 5, 10, 9, 24],
  [1992, 5, 3, 12, 20],
  [1996, 1, 8, 6, 15],
  [2000, 11, 10, 7, 28],
  [2004, 9, 9, 12, 30],
  [2008, 19, 13, 19, 51],
  [2012, 29, 17, 19, 65],
  [2016, 27, 23, 17, 67],
  [2020, 22, 21, 22, 65]],
 'HUN': [[1896, 2, 1, 3, 6],
  [1900, 1, 2, 2, 5],
  [1904, 2, 1, 1, 4],
  [1908, 3, 4, 2, 9],
  [1912, 3, 2, 3, 8],
  [1924, 2, 3, 4, 9],
  [1928, 4, 5, 0, 9],
  [1932, 6, 4, 5, 15],
  [1936, 10, 1, 5, 16],
  [1948, 10, 5, 12, 27],
  [1952, 16, 10, 16, 42],
  [1956, 9, 10, 7, 26],
  [1960

In [70]:
[country_codes[code] for code in 
 increasing_medals(participated_l_5(data_totals_dict)).keys()]

['Belgium', 'New Zealand', 'Japan']

In [69]:
increasing_medals(participated_l_5(data_totals_dict))

{'BEL': [[1900, 6, 7, 4, 17],
  [1908, 1, 5, 2, 8],
  [1912, 2, 1, 3, 6],
  [1920, 16, 12, 14, 42],
  [1924, 3, 7, 3, 13],
  [1928, 0, 1, 2, 3],
  [1936, 0, 0, 2, 2],
  [1948, 2, 2, 3, 7],
  [1952, 2, 2, 0, 4],
  [1956, 0, 2, 0, 2],
  [1960, 0, 2, 2, 4],
  [1964, 2, 0, 1, 3],
  [1968, 0, 1, 1, 2],
  [1972, 0, 2, 0, 2],
  [1976, 0, 3, 3, 6],
  [1980, 1, 0, 0, 1],
  [1984, 1, 1, 2, 4],
  [1988, 0, 0, 2, 2],
  [1992, 0, 1, 2, 3],
  [1996, 2, 2, 2, 6],
  [2000, 0, 2, 3, 5],
  [2004, 1, 0, 2, 3],
  [2008, 2, 0, 0, 2],
  [2012, 0, 1, 2, 3],
  [2016, 2, 2, 2, 6],
  [2020, 3, 1, 3, 7]],
 'NZL': [[1920, 0, 0, 1, 1],
  [1924, 0, 0, 1, 1],
  [1928, 1, 0, 0, 1],
  [1932, 0, 1, 0, 1],
  [1936, 1, 0, 0, 1],
  [1952, 1, 0, 2, 3],
  [1956, 2, 0, 0, 2],
  [1960, 2, 0, 1, 3],
  [1964, 3, 0, 2, 5],
  [1968, 1, 0, 2, 3],
  [1972, 1, 1, 1, 3],
  [1976, 2, 1, 1, 4],
  [1984, 8, 1, 2, 11],
  [1988, 3, 2, 8, 13],
  [1992, 1, 4, 5, 10],
  [1996, 3, 2, 1, 6],
  [2000, 1, 0, 3, 4],
  [2004, 3, 2, 0, 5],
  [2008,

New Zealand is the only country that increased its medal count over the last four olympics, going up from 5 in 2004 to 9, 13, and a finally a very impressive 20 in 2020.  Belgium and Japan both increased their medal counts over the last three Olympics.

###### 10b Decreased over last four olympics

I am intrepreting this to mean 2020<2016<2012<2008<2004.

In [72]:
def decreasing_medals(d = dict()):
    '''return a dictionary including the items for which there is an increase
    in total medals for each of the last four olympics'''
    return {c: y_m for c, y_m in d.items() if \
     (y_m[-1][4] < y_m[-2][4])&\
     (y_m[-2][4] < y_m[-3][4])&\
     (y_m[-3][4] < y_m[-4][4])\
     &(y_m[-4][4] < y_m[-5][4])
    }

In [73]:
[country_codes[code] for code in 
 decreasing_medals(participated_l_5(data_totals_dict)).keys()]

[]

No country decreased its medals count over each of the last four Olympics. South Korea and Belarus decreased their medal counts over the last three Olympics.